internal/backend/telemetry/telemetry.go

```diff
@@ -0,0 +1,140 @@
+	ServiceName string `koanf:"service_name"`
+	Endpoint    string `koanf:"endpoint"`
+}
+
+var Configs = configset.Set[Config]{
+	Default: &Config{Enabled: true, ServiceName: "ak", Endpoint: "localhost:4318"},
+	Dev:     &Config{Enabled: false, ServiceName: "ak", Endpoint: "localhost:4318"},
+}
+
+func verifyConfig(cfg *Config) *Config {
+	if cfg == nil {
+		return Configs.Default
+	}
+	if cfg.ServiceName == "" {
+		cfg.ServiceName = Configs.Default.ServiceName
+	}
+	if cfg.Endpoint == "" {
+		cfg.Endpoint = Configs.Default.Endpoint
+	}
+	return cfg
+}
+
+type Labels map[string]string
+
+func WithLabels(labels Labels) api.MeasurementOption {
+	var attrs []attribute.KeyValue
+	for k, v := range labels {
+		attrs = append(attrs, attribute.String(k, v))
+	}
+	return api.WithAttributes(attrs...)
+}
+
+type Telemetry struct {
+	z           *zap.Logger
+	enabled     bool
+	serviceName string
+}
+
+func New(z *zap.Logger, cfg *Config) *Telemetry {
+	cfg = verifyConfig(cfg)
+
+	telemetry := &Telemetry{z: z, enabled: cfg.Enabled, serviceName: cfg.ServiceName}
+
+	if !telemetry.enabled {
+		z.Info("metrics are disabled")
+		return telemetry
+	}
+
+	// REVIEW: encryption? GRPC?
+	// REVIEW: should we use AK env or maybe standard OTEL onses?
+	//         - OTEL_EXPORTER_OTLP_ENDPOINT or OTEL_EXPORTER_OTLP_METRICS_ENDPOINT
+	exporter, err := otlpmetrichttp.New(
+		context.Background(),
+		otlpmetrichttp.WithInsecure(),
+		otlpmetrichttp.WithEndpoint(cfg.Endpoint),
+		// otlpmetrichttp.WithURLPath("/"), // /v1/metrics by default, unless set
+	)
+	if err != nil {
+		z.Error("failed to create metric exporter: %v", zap.Error(err))
+		telemetry.enabled = false
+		return telemetry
+	}
+
+	schemaURL := "https://opentelemetry.io/schemas/1.1.0"
+	resourceAttrs := resource.NewWithAttributes(
+		schemaURL,
+		semconv.ServiceNameKey.String(telemetry.serviceName),
+	)
+
+	// REVIEW: consider using controller?
+	meterProvider := metric.NewMeterProvider(
+		metric.WithReader(metric.NewPeriodicReader(exporter)),
+		metric.WithResource(resourceAttrs),
+	)
+
+	otel.SetMeterProvider(meterProvider) // set global meter provider
+	return telemetry
+}
+
+type NoOpMetric struct {
+	api.Int64UpDownCounter
+	api.Int64Counter
+}
+
+func (NoOpMetric) Add(context.Context, int64, ...api.AddOption) {}
+
+func newMetric[T any](t *Telemetry, name string, description string,
+	createFunc func(meter api.Meter, name string, description string) (T, error),
+) T {
+	meter := otel.GetMeterProvider().Meter(t.serviceName)
+	if !strings.HasPrefix(name, t.serviceName) {
+		name = fmt.Sprintf("%s.%s", t.serviceName, name)
+	}
+	metric, err := createFunc(meter, name, description)
+	if err != nil {
+		t.z.Error("failed to create metric", zap.String("name", name), zap.Error(err))
+		// REVIEW: should we panic? kittehs.Must?
+	}
+	return metric
+}
+
+func (t *Telemetry) NewUpDownCounter(name string, description string) api.Int64UpDownCounter {
+	if !t.enabled {
+		return NoOpMetric{}
```

Comment:
if we don't have real telemetry installed (meaning no telemetry initialized), what will `Int64UpDownCounter` do? Maybe it's just a no-op without special help?

Reply:
rewritten to use opentelemetry no-ops.

Reply:
rewritten to use opentelemetry no-ops.